File: compliance/python/workflows/order_processing.py
"""
Order Processing Compliance Workflow (Python reference implementation).

Multi-step workflow: validate order → charge payment (with retries) → send confirmation.
Outputs the expected operation sequence as JSON for Rust compliance verification.

Usage:
    python order_processing.py > ../tests/fixtures/order_processing.json
"""

import json


def generate_expected_operations():
    """Generate the operation sequence that this workflow produces.

    The Python durable Lambda SDK records each operation as:
    - type: "step" for ctx.step() calls
    - name: the operation name string

    This workflow performs three sequential steps:
    1. validate_order — validates the incoming order
    2. charge_payment — charges payment (configured with retries)
    3. send_confirmation — sends order confirmation
    """
    return {
        "workflow": "order_processing",
        "operations": [
            {"type": "step", "name": "validate_order"},
            {"type": "step", "name": "charge_payment"},
            {"type": "step", "name": "send_confirmation"},
        ],
    }


if __name__ == "__main__":
    print(json.dumps(generate_expected_operations(), indent=2))

File: compliance/rust/src/order_processing.rs
//! Order Processing compliance workflow.
//!
//! Multi-step workflow: validate order → charge payment (with retries) → send confirmation.
//! Mirrors the Python reference implementation in `compliance/python/workflows/order_processing.py`.

use durable_lambda_core::context::DurableContext;
use durable_lambda_core::error::DurableError;
use durable_lambda_core::types::StepOptions;

/// Execute the order processing workflow against the given context.
///
/// Operations recorded:
/// 1. `step:validate_order`
/// 2. `step:charge_payment`
/// 3. `step:send_confirmation`
pub async fn run(ctx: &mut DurableContext) -> Result<serde_json::Value, DurableError> {
    // Step 1: Validate the order
    let validated: Result<serde_json::Value, String> = ctx
        .step("validate_order", || async {
            Ok(serde_json::json!({"order_id": "ORD-001", "valid": true}))
        })
        .await?;

    // Step 2: Charge payment with retries
    let charged: Result<serde_json::Value, String> = ctx
        .step_with_options(
            "charge_payment",
            StepOptions::new().retries(3).backoff_seconds(2),
            || async { Ok(serde_json::json!({"charged": true, "amount": 99.99})) },
        )
        .await?;

    // Step 3: Send confirmation
    let confirmed: Result<serde_json::Value, String> = ctx
        .step("send_confirmation", || async {
            Ok(serde_json::json!({"confirmed": true, "email_sent": true}))
        })
        .await?;

    Ok(serde_json::json!({
        "validated": validated.unwrap(),
        "charged": charged.unwrap(),
        "confirmed": confirmed.unwrap(),
    }))
}
